08_project/runnerdave/pkg/puppy/store/sync_store.go

```diff
@@ -0,0 +1,64 @@
+package store
+
+import (
+	"sync"
+
+	puppy "github.com/anz-bank/go-course/08_project/runnerdave/pkg/puppy"
+)
+
+// SyncStore implementation of the Storer interface
+type SyncStore struct {
+	sync.Mutex
+	sync.Map
+}
+
+// NewSyncStore creates a new in-memory store with map intialised
+func NewSyncStore() *SyncStore {
+	return &SyncStore{}
+}
+
+// CreatePuppy saves new puppy if not in store, if it is already returns error
+func (s *SyncStore) CreatePuppy(p *puppy.Puppy) error {
+	if err := puppy.ValidateValue(p.Value); err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.Load(p.ID); ok {
+		return puppy.Errorf(puppy.ErrUnknown, "puppy with id %d already exists", p.ID)
+	}
+	s.Store(p.ID, *p)
+	return nil
+}
+
+// ReadPuppy gets a puppy from the store given an ID
+func (s *SyncStore) ReadPuppy(id uint16) (puppy.Puppy, error) {
+	if puppyData, ok := s.Load(id); ok {
+		puppy, _ := puppyData.(puppy.Puppy)
+		return puppy, nil
+	}
+	return puppy.Puppy{}, puppy.Errorf(puppy.ErrIDNotFound, "puppy with ID:%d not found", id)
+}
+
+// UpdatePuppy puts new puppy data to the store, error if id does not exist
+func (s *SyncStore) UpdatePuppy(id uint16, p *puppy.Puppy) error {
+	if err := puppy.ValidateValue(p.Value); err != nil {
+		return err
+	}
+	if _, ok := s.Load(id); !ok {
```

Comment:
sorry,  I missed this before -   think you should add the mutex here too.

Reply:
Sorry @juliaogris you mean to add a lock? I don't understand

Reply:
yes, apologies for the confusion you load and you store - this should happen as an atomic action, so please add
```go
         s.Lock()
	defer s.Unlock()
```
as below.

Reply:
I see, thanks for pointing that out, implemented